### PR TITLE
Moddable Minimap

### PR DESCRIPTION
**Pull request by est31**
This commit allows most aspects of the minimap to be
controlled from server side LUA, while still allowing
for lag free control of the minimap via client keyboard.

The server can upload a state machine of their desire
to the client.

The network protocol format is kept extensible to
allow for additional fields without protocol bumps.

Thanks to the design the server can take over full
control and use single-state state machines to change
every aspect of the minimap without ways to intervene
from the client (expect for disabling minimap altogether).

This allows for "map" mods that only display the minimap
if you wield a "map" craftitem.

This is still very WIP and in early development stage.

TODO: see commit msg, but lots of stuff

### CODE REVIEW

File: src/script/lua_api/l_object.cpp [hunk 1]
```diff
@@ -1181,6 +1181,29 @@ int ObjectRef::l_get_player_control_bits(lua_State *L)
...
+// set_minimap_mode(self, id)
+int ObjectRef::l_set_minimap_mode(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	Player *player = getplayer(ref);
+	if (player == NULL)
+		return 0;
+	// TODO
+}
```
Comment (Gael-de-Sailly) [inline src/script/lua_api/l_object.cpp]: I have a warning during compilation: `warning: control reaches end of non-void function [-Wreturn-type]`

File: src/script/lua_api/l_object.cpp [hunk 2]
```diff
@@ -1181,6 +1181,29 @@ int ObjectRef::l_get_player_control_bits(lua_State *L)
...
+// set_minimap_modes(self, modes)
+int ObjectRef::l_set_minimap_modes(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	Player *player = getplayer(ref);
+	if (player == NULL)
+		return 0;
+	// TODO
+}
```
Comment (Gael-de-Sailly) [inline src/script/lua_api/l_object.cpp]: and here too

File: src/hud.cpp [hunk 3]
```diff
@@ -39,6 +43,81 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+}
+
+MinimapModeMachine::~MinimapModeMachine()
+{
+	delete[] m_defs;
+}
+
+void MinimapModeMachine::serialize(std::ostream &os)
+{
+	for (u16 i; i < m_mode_count; i++) {
```
Comment (Gael-de-Sailly) [inline src/hud.cpp]: Another warning here: `warning: ‘i’ may be used uninitialized in this function [-Wmaybe-uninitialized]`

File: src/hud.cpp [hunk 4]
```diff
@@ -39,6 +43,81 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+		writeU16(oss, m.next_mode_shift);
+		// Add future fields here
+
+		os << serializeString(oss.str());
+	}
+}
+
+void MinimapModeMachine::deserialize(std::istream &is)
+{
+	for (u16 i; i < m_mode_count; i++) {
```
Comment (Gael-de-Sailly) [inline src/hud.cpp]: and here too

File: src/script/lua_api/l_object.cpp (hunk 1 above)
Comment (est31) [inline src/script/lua_api/l_object.cpp]: well yeah, the PR is not complete

File: src/script/lua_api/l_object.cpp (hunk 2 above)
Comment (Gael-de-Sailly) [inline src/script/lua_api/l_object.cpp]: Ok. Sorry for having read too fast :)

**Comment by paramat** [comment]
est31 i'll remove this from milestone as you said you probably can't finish it in time for release, hope this is ok.

**Comment by 0-afflatus** [comment]
Definitely the WTG, deserves time to do it right.

**Comment by paramat** [comment]
@est31 another relaease coming up, this would be good to have.

**Comment by est31** [comment]
Due to CSM I think this PR is obsolete, at least in the current form. Therefore, closing.

**Comment by tacotexmex** [comment]
> Thanks to the design the server can take over full
control and use single-state state machines to change
every aspect of the minimap without ways to intervene
from the client (expect for disabling minimap altogether).

How is that related to CSM? CSM mods aren't controlled by server, are they?

**Comment by est31** [comment]
@tacotexmex I was assuming a non hacked client in that statement. For minimaps, you can't protect from hacked clients.

**Comment by tacotexmex** [comment]
It's still nice to be able to build mods around it though, as mentioned in #4378

**Pull request closed without merging**